### ISSUE TITLE
Build/Compile "next" builds and provide these builds for download

### ISSUE THREAD

**Issue by JohnArcher** [htacg/tidy-html5#587]
I would like to use (and test) the latest version, but I can't compile the source for different reasons. Is it possible that you could provide the builds of the latest version, marked as beta/experimental/... for download, e.g. on http://binaries.html-tidy.org or here on github?

I don't know whether you could automate that, but I think it would be ok to provide a new version once a week, if you have to do it manually.

(Additional, personal background: I would like the fix mentioned in #461 and fixed in #540 ...)

Thanks a lot!

**Comment by balthisar** [comment]
There are a lot of targets, so building all of them every week would be quite time consuming. I can build you something for macOS if you like.

**Comment by JohnArcher** [comment]
Ok, thanks for the info. Well, I would need a version for Windows, though, but thanks for the offer!

**Comment by geoffmcl** [comment]
@JohnArcher as you can read in several other posts, I am **all for** trying to do this... and have done it for certain versions, to make it easy for others to get and try a specific version, now and again... but...

It becomes more **where** to put them? And how and when?

The obvious place might seem to be [Releases](https://github.com/htacg/tidy-html5/releases)... you will note at present there are **seven**, and one of these is actually a **next** marked as `Pre-release` 5.1.14...

But this has the **BIG** problem that it requires adding a **tag**, but I wanted to keep **tags** `only` for `official` releases, like we have `5.0.0`, `5.2.0`, `5.4.0`, and hopefully `5.6.0` soon - **note** the **even** numbers...

I suppose ideally we could create a `Pre-release` tag every time we **bump** the version number, but this has big problems that (1) we would have to remember to go back and delete older `next` tags, otherwise it seems we would get a too big, confusing list, and (2) this creating a tag, setting up a `pre-release`, and adding binaries, is a **big** manual task... a drag and drop process...

Maybe this can be achieved through the github api? Do not know... any pointers?

And just checking, it seems it is also quite a number of **manual steps** in doing the deletes...

While it is agreed some of this could be semi-automated via scripts - I would try to look after a Windows `next` release zip, maybe both 64 and 32 bit, and maybe a unix Ubuntu deb and rpm, and @balthisar with  something for the macOS... 

But as he points out, building them for every version bump, which can be each week, or even more frequently in some cases, would be quite **time consuming**... quite an extra burden...

So while I started out being all for this, I am sort of talking myself out of it, due the **big extra manual effort** involved...

Thus it feels like the `github.com` release system is **too much effort** each time for `next`...

It is agreed, maybe we should try to increase the official release cycle... I did want to try for at least twice a year, but have not been able to achieve that...

**Are there other ideas?**

As a developer, I have over 150 Windows versions available, so I can easily do backward testing... like -

```
tidy-2000.exe                     tidy-2004.exe                     tidy-2009.exe
... etc ...
tidy5-5.5.13.exe                  tidy5-5.5.14.exe                  tidy5-5.5.15.exe
tidy5-5.5.16.exe                  tidy5-5.5.17.exe                  tidy5-5.5.18.exe
tidy5-5.5.2.exe                   tidy5-5.5.20.exe                  tidy5-5.5.21.exe
tidy5-5.5.24.exe                  tidy5-5.5.27.exe                  tidy5-5.5.31-1.exe
tidy5-5.5.31.exe                  tidy5-5.5.31.I567.exe             tidy5-5.5.31.w32-vc14-mt-xp.exe
```

Could I just maybe zip these and copy them to say my [site](http://geoffair.org/tidy/), like I used to do **years** ago? Last was sort of 2012... 

Is there another site, that could be used?

The idea would be to add a little to my build scripts, and as I build each `new` version, it is like `scp'ed`, or some other `copy` tool, to the site... all done...

Of course the other idea is that people should be encouraged to **build tidy from source**... this has long been the `tradition` in linux, and given tidy's very minimal dependance is quite easy to achieve in Windows using `free` MSVC, and other tools... not sure what is available in the macOS... maybe it is **asking too much** that the current maintainer try to take on this task...

Anyway, I am just throwing out ideas here, and seek feedback... I will certainly do something if it is **easy**... thanks...

**Comment by balthisar** [comment]
We *could* go back to one of the continuous integration solutions; I don't really like these, but something like Travis can be configured to automatically post builds to one of our gh_pages. The issue with this, though, is that it can only generate bins for macOS and for Ubuntu. _If_ someone wants to develop a cross-compilation technique via macOS or Linux that includes Windows (required) and other platforms (would be nice), then _maybe_ we could reconsider Travis-CI or something similar.

I will _not_ consider an automated solution that leaves out a huge, huge Tidy-using population, though, and so Windows cross-compilation would be an un-negotiable requirement.

**Comment by geoffmcl** [comment]
@balthisar agree with all you have said, including I too really do not like such CI automation... even though some rave about it... very hard to put my finger exactly on the problems I see... but, as replied to @JohnArcher, would like to do something **easy**...

Concerning Windows it seems github does offer [AppVeyor](https://github.com/marketplace/appveyor), which is **free** for open source projects, which tidy is...

Seems someone, but not me, needs to **test and experiment**... my concerns and questions would be -

  - Is this MSVC Win64 only? Which version? What about 32-bits? What about special builds, like static `/MT'?
  - Exactly **where** would the binaries be posted? I do **not** think this should be `htacg/tidy-html5` repo.
  - What about MingGW build? 64 and 32 bits?
  - And then do we have to also have Travis CI for `unix`?
  - What about the `macOS`? Others?
  - And probably lots **more** questions if this progresses...

I can certainly say this CI process concerns me... and as stated, it should cover **multiple** OSes... Will **not** accept just one, or 2... min. 3 - unix, macOS, win32, but hopefully more...

But if this `AppVeyor` can build and deploy binaries to say a `gh_pages` site, why can't I set that up here?

Why are we using a 3rd party to do the building, when the `scp` copying after a build here is just as easy for me... it would probably be just a few lines added to the batch files I already use... and is much more under my control... 

So the single main problem for me is the **target** deployment site... can someone add some thought, details on that... thanks...

I would see this 2nd binary site setup is required, whatever we eventually decide to do... I certainly would not want http://binaries.html-tidy.org/ cluttered with, in this case some 38 `next` binaries... tidy-5.5.1 through to tidy-5.5.38, etc... until the next release...

In fact maybe the current `binaries.html-tidy.org` could be changed to `releases.html-tidy.org`, and `binaries` replaced with say `next.html-tidy.org`, but have no idea how to do that...

As stated, just ideas, and would go along with anything **easy**...

**Comment by JohnArcher** [comment]
(Just want to say, I read you comments and appreciate them a lot! I have no skills in this field, but I try to have a look at other projects how they handle that. But I am aware of the problem you have to deal with as you support so many platforms (which is cool!)).

**Comment by balthisar** [comment]
I wonder what is a population of users that want to use `next` that are not able to build it themselves? End users aren't what they once were; installing a compiler and CMake _can_ be quite scary to someone who's a humble HTML developer, and in workplaces group policy may even make this impossible. I have no data, but can only suspect that there's probably a valid case for providing `next` binaries from time to time.

But... do we want to do this for every single commit? If not, then I don't mind building and posting manually from time to time. If we want to do it for every single commit, then this really has be to automated.

> So the single main problem for me is the target deployment site... can someone add some thought, details on that... thanks...

It looks like these both CI solutions can dump bins into a gh-pages repository, but I've only looked at it on the surface. We could set up a repo just for "daily builds," but Github doesn't really like us distributing binaries from gh-pages, thus why we re-worked our binaries page recently to link to releases rather than a directory in gh-pages.

I would only want to link to the current `next`; there's no good reason to keep a binary for every commit; people who need a specific commit can build that commit by themselves. Maybe github wouldn't be upset if we hosted only a single binary, except, of course, every binary ultimately would still take space as once they're in the repo, they're permanently in the repo's history.

I'm not going to invest any real time in looking at this and getting it to work unless there's some real interest in this, though.
